artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/PriorityCollection.java

```diff
@@ -0,0 +1,319 @@
+   }
+
+   @Override
+   public Iterator<T> iterator() {
+      Iterator<T>[] iterators = getIterators();
+      return new MultiIterator<>(iterators);
+   }
+
+   private Iterator<T>[] getIterators() {
+      PriorityHolder<T>[] snapshot = this.getArray();
+      int size = snapshot.length;
+      Iterator<T>[] iterators = newIteratorArrayInstance(size);
+      for (int i = 0; i < size; i++) {
+         iterators[i] = snapshot[i].getValues().iterator();
+      }
+      return iterators;
+   }
+
+   @SuppressWarnings("unchecked")
+   private static <T> Iterator<T>[] newIteratorArrayInstance(int length) {
+      return (Iterator<T>[]) Array.newInstance(Iterator.class, length);
+   }
+
+   public ResettableIterator<T> resettableIterator() {
+      return new MultiResettableIterator<T>(getResettableIterators());
+   }
+
+   private ResettableIterator<T>[] getResettableIterators() {
+      PriorityHolder<T>[] snapshot = this.getArray();
+      int size = snapshot.length;
+      ResettableIterator<T>[] iterators = newResettableIteratorArrayInstance(size);
+      for (int i = 0; i < size; i++) {
+         iterators[i] = new ArrayResettableIterator<>(snapshot[i].getValues().toArray());
+      }
+      return iterators;
+   }
+
+   @SuppressWarnings("unchecked")
+   private static <T> ResettableIterator<T>[] newResettableIteratorArrayInstance(int length) {
+      return (ResettableIterator<T>[]) Array.newInstance(ResettableIterator.class, length);
+   }
+
+   @Override
+   public void forEach(Consumer<? super T> action) {
+      Objects.requireNonNull(action);
+      PriorityHolder<T>[] current = getArray();
+      int len = current.length;
+      for (int i = 0; i < len; ++i) {
+         current[i].getValues().forEach(action);
+      }
+   }
+
+   private Collection<T> getCollection(int priority, boolean createIfMissing) {
+      PriorityHolder<T>[] current = getArray();
+      int low = 0;
+      int high = current.length - 1;
+
+      while (low <= high) {
+         int mid = (low + high) >>> 1;
+         PriorityHolder<T> midVal = current[mid];
+
+         if (midVal.getPriority() > priority)
+            low = mid + 1;
+         else if (midVal.getPriority() < priority)
+            high = mid - 1;
+         else
+            return midVal.getValues(); //key found
+      }
+
+      if (createIfMissing) {
+         PriorityHolder<T>[] newArray = newPrioritySetArrayInstance(current.length + 1);
+         if (low > 0) {
+            System.arraycopy(current, 0, newArray, 0, low);
+         }
+         if (current.length - low > 0) {
+            System.arraycopy(current, low, newArray, low + 1, current.length - low);
+         }
+         newArray[low] = new PriorityHolder<T>(priority, supplier);
+         setArray(newArray);
+         return newArray[low].getValues();
+      }
+      return null;
+   }
+
+   @Override
+   public synchronized boolean add(T t) {
```

Comment:
Looking to the usage on `QueueImpl` it seems that is being always called into a `synchronized (this) {` block ie it doesn't need to be `synchronized` and same same for `remove`. 
If we allow `snapshotting` from a different thread is fine, but I would document it instead.
I know that's a less safe approach than putting `sychronized` all around, but I do believe that we would have to do the same on  `LinkedListImpl` for this same reason if we need to `protect` correct usage at any cost.

Reply:
This needs to be concurrent safe on modifications